### PR TITLE
use generic type to make code cleaner

### PR DESCRIPTION
**Pull request by XavierM**
When I was reviewing you PR, I wanted to make sure that not all props where required to make the component work. So I came up with this solution.

Let me know, if you like it too

### CODE REVIEW

File: x-pack/plugins/triggers_actions_ui/public/application/sections/rules_list/components/rule_tag_badge.tsx [hunk 1]
```diff
@@ -16,55 +16,61 @@ const tagTitle = i18n.translate(
...
+export type RuleTagBadgeOptions = 'all' | 'default';
+
+interface RuleTagBadgeBasicOptions {
+  isOpen: boolean;
+  onClick: React.MouseEventHandler<HTMLButtonElement>;
+  onClose: () => void;
+}
+
 export interface RuleTagBadgeCommonProps {
+  showAllTags?: boolean;
```
Comment (fkanout) [inline x-pack/plugins/triggers_actions_ui/public/application/sections/rules_list/components/rule_tag_badge.tsx:28]: nit: I understand it as if I set `showAllTags` to `false`, some tags will be hidden for an unknown reason. So maybe something like `expandTags`, `spreadTags`, `unfoldTags` would be a better choice.

File: x-pack/plugins/triggers_actions_ui/public/application/sections/rules_list/components/rule_tag_badge.tsx [hunk 2]
```diff
@@ -16,55 +16,61 @@ const tagTitle = i18n.translate(
...
-      isOpen: boolean;
-      onClick: React.MouseEventHandler<HTMLButtonElement>;
-      onClose: () => void;
-    }
-  | {
-      spread: boolean;
-    };
-export type RuleTagBadgeProps = RuleTagBadgeCommonProps & RuleTagBadgeConditionalProps;
+
+export type RuleTagBadgeProps<T extends RuleTagBadgeOptions = 'default'> = T extends 'default'
```
Comment (fkanout) [inline x-pack/plugins/triggers_actions_ui/public/application/sections/rules_list/components/rule_tag_badge.tsx:35]: For me `RuleTagBadgeProps ` is bit complicated, with these types `RuleTagBadgeOptions`,  `RuleTagBadgeCommonProps`,  and `RuleTagBadgeBasicOptions`.

File: x-pack/plugins/triggers_actions_ui/public/application/sections/rules_list/components/rule_tag_badge.tsx [hunk 3]
```diff
@@ -16,55 +16,61 @@ const tagTitle = i18n.translate(
   }
 );
 
+export type RuleTagBadgeOptions = 'all' | 'default';
```
Comment (fkanout) [inline x-pack/plugins/triggers_actions_ui/public/application/sections/rules_list/components/rule_tag_badge.tsx:19]: I wanted this update to be invisible (no update required after this change) for any component/page uses`RuleTag`. But does introduce the `default` option will require updating all component uses tags? 

nit: `all` is very generic and doesn't reflect the functionality.

**Review by fkanout — COMMENTED** [review]
@XavierM, many thanks for the time invested in proposing this solution. I liked the approach of using the type utility, I will try to update my PR based on that. Then we can decide how to proceed.

File: x-pack/plugins/triggers_actions_ui/public/application/sections/rules_list/components/rule_tag_badge.tsx (hunk 1 above)
Comment (XavierM) [reply x-pack/plugins/triggers_actions_ui/public/application/sections/rules_list/components/rule_tag_badge.tsx:28]: So maybe, we can do the other way and create `tagInPopover` as a boolean, I think that might be simpler

Comment (fkanout) [reply x-pack/plugins/triggers_actions_ui/public/application/sections/rules_list/components/rule_tag_badge.tsx:28]: Perfect, I like it 👍🏻 !

**Review by fkanout — APPROVED** [review]
I will merge it to my branch, then just some tweaks there and update some wording. Thanks @XavierM !

**Pull request merged**